src/main/webapp/app/course/learning-paths/participate/learning-path-storage.service.ts

```diff
@@ -0,0 +1,211 @@
+     * @param entry the entry that should be set to interacted with
+     * @param value the value that should be set
+     */
+    setInteraction(learningPathId: number, entry: StorageEntry, value: boolean) {
+        if (!this.learningPathRecommendations.has(learningPathId)) {
+            return;
+        }
+        const storedEntry = this.getStoredRecommendationEntry(learningPathId, entry);
+        if (storedEntry) {
+            storedEntry.interacted = value;
+        }
+    }
+
+    /**
+     * Gets the next recommended entry for a learning object.
+     * <p>
+     * If the given entry has no successor, the first entry that has not been interacted with will be returned.
+     * @param learningPathId the id of the learning path
+     * @param entry the entry for which the successor should be returned
+     */
+    getNextRecommendation(learningPathId: number, entry: StorageEntry | undefined): StorageEntry | undefined {
+        if (!this.learningPathRecommendations.has(learningPathId)) {
+            return undefined;
+        }
+        // if no entry given, retrieve first not interacted entry
+        if (!entry) {
+            const nextEntry = this.learningPathRecommendations.get(learningPathId)!.find((e) => !e.interacted);
+            if (nextEntry) {
+                nextEntry.interacted = true;
+            }
+            return nextEntry;
+        }
+        const storedEntry = this.getStoredRecommendationEntry(learningPathId, entry);
+        const nextIndex = this.learningPathRecommendations.get(learningPathId)!.indexOf(storedEntry!) + 1;
+        const nextEntry = this.learningPathRecommendations.get(learningPathId)!.find((e, idx) => idx >= nextIndex && !e.interacted);
+        if (nextEntry) {
+            // if there is a successor, update interaction and return
+            nextEntry.interacted = true;
+            return nextEntry;
+        } else {
+            // if there is no successor, retrieve first not interacted entry
+            return this.getNextRecommendation(learningPathId, undefined);
+        }
+    }
+
+    private getStoredRecommendationEntry(learningPathId: number, entry: StorageEntry) {
+        return this.learningPathRecommendations.get(learningPathId)!.find((e) => {
+            if (e instanceof LectureUnitEntry && entry instanceof LectureUnitEntry) {
+                return e.lectureId === entry.lectureId && e.lectureUnitId === entry.lectureUnitId;
+            } else if (e instanceof ExerciseEntry && entry instanceof ExerciseEntry) {
+                return e.exerciseId === entry.exerciseId;
+            }
+            return false;
+        });
+    }
+}
+
+export abstract class StorageEntry {
+    interacted = false;
+
+    abstract equals(other: StorageEntry): boolean;
+}
+
+export class LectureUnitEntry extends StorageEntry {
+    readonly lectureUnitId: number;
+    readonly lectureId: number;
+
+    constructor(lectureId: number, lectureUnitId: number) {
+        super();
```

Comment:
Can you leave this out? At least in Java this call would be optional

Reply:
No, the IDE and compiler immediately complain ^^